crates/starknet-devnet-server/src/api/json_rpc/endpoints.rs

```diff
@@ -23,14 +23,20 @@ impl JsonRpcHandler {
     }
 
     /// starknet_getBlockWithTxHashes
-    pub async fn get_block_with_tx_hashes(&self, block_id: BlockId) -> StrictRpcResult {
-        let block =
-            self.api.starknet.read().await.get_block(block_id.as_ref()).map_err(
-                |err| match err {
-                    Error::NoBlock => ApiError::BlockNotFound,
-                    unknown_error => ApiError::StarknetDevnetError(unknown_error),
-                },
-            )?;
+    pub async fn get_block_with_tx_hashes(&self, mut block_id: BlockId) -> StrictRpcResult {
+        let starknet = self.api.starknet.read().await;
+
+        if !starknet.config.blocks_on_demand
+            && block_id == BlockId(ImportedBlockId::Tag(BlockTag::Pending))
+        {
+            // in normal mode, querying pending block should default to the latest
+            block_id = BlockId(ImportedBlockId::Tag(BlockTag::Latest));
+        }
```

Comment:
It is easier to have a conversation in comments than in posts on PR. Continuing from https://github.com/0xSpaceShard/starknet-devnet-rs/pull/430#issuecomment-2076689281

I feel like 35be702cba97a1e8f41fbf306f271de875722239 only partly addresses the original issue. Shouldn't we always default to latest if no pending, for all rpc calls? Also, I feel like this block selection should be done as deep as possible, e.g. in `get_by_block_id`

Reply:
In `Starknet` struct there is a `restart_pending_block()` which is resetting `self.blocks.pending_block` so pending_block is always created and updated so there is no way of detecting if there is one or not.

And about `get_by_block_id()` - in this layer we don't have information about `starknet.config.blocks_on_demand` because it's inside`StarknetBlocks`.

Reply:
Would our problems be solved if pending block became `Option`?

Reply:
probably yes but this would require getting rid of `restart_pending_block()` logic and it's not the scope of this ticket

Reply:
Making block_id=`pending` default to `latest` is the scope of this ticket, so we're in a dilemma

Reply:
The current code is defaulting to the `latest`, the problem is that the current business logic is using `restart_pending_block()` and this should be removed.

What I may suggest is to open a new ticket with the refactoring of restart_pending_block() and making `self.blocks.pending_block` optional.

Reply:
https://github.com/0xSpaceShard/starknet-devnet-rs/issues/445

Reply:
I'll check that

Reply:
fixed in 16004864730e942e84f8a0371b8f44ec989a717a, is not that bad with blocks_on_demand inside StarknetBlocks, actually it makes sense